File: src/main/java/pers/neige/neigeitems/action/node/impl/JavascriptParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.manager.ScriptManager;
import pers.neige.neigeitems.utils.StringUtils;

import java.util.HashMap;
import java.util.List;
import java.util.function.Function;

import static pers.neige.neigeitems.manager.HookerManager.papi;

public class JavascriptParser extends NodeParser {
    public JavascriptParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "js";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            context,
            getParsedValue(context, params, "path"),
            params.getStringList("args")
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val paramsList = StringUtils.split(params, '_', '\\');
        val path = paramsList.get(0);
        paramsList.remove(0);
        return handle(
            context,
            path,
            paramsList
        );
    }

    private @Nullable String handle(
        @NonNull ActionContext context,
        @Nullable String info,
        @NonNull List<?> args
    ) {
        if (info == null) {
            warning("未指定脚本路径");
            return null;
        }
        val array = info.split("::", 2);
        val path = array[0];
        val script = ScriptManager.INSTANCE.getCompiledScripts().get(path);
        if (script == null) {
            warning("不存在脚本 " + path);
            return null;
        }
        val func = array[1];
        val map = new HashMap<String, Object>();
        if (context.getCaster() instanceof OfflinePlayer) {
            map.put("player", context.getCaster());
            map.put("papi", (Function<String, String>) (string) -> papi((OfflinePlayer) context.getCaster(), string));
        }
        map.put("vars", (Function<String, String>) (string) -> this.manager.parseNode(string, context));
        return this.manager.parseNode(String.valueOf(script.invoke(func, map, args.toArray())), context);
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/JavascriptParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.manager.HookerManager.papi
import pers.neige.neigeitems.manager.ScriptManager
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.SectionUtils.parseSection

/**
 * js节点解析器
 */
object JavascriptParser : SectionParser() {
    override val id: String = "js"

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String? {
        return handler(
            cache,
            player,
            sections,
            true,
            data.getString("path"),
            data.getStringList("args")
        )
    }

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(
            cache,
            player,
            sections,
            false,
            args.getOrNull(0),
            if (args.isNotEmpty()) args.drop(1) else mutableListOf()
        ) ?: "<$id::${args.joinToString("_")}>"
    }


    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param parse 是否对参数进行节点解析
     * @param info 脚本文件名::函数名
     * @param args 函数参数
     * @return 解析值
     */
    private fun handler(
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        parse: Boolean,
        info: String?,
        args: List<String>
    ): String? {
        info?.let {
            val array = info.split("::")
            // 脚本文件名
            val path = array[0]
            // 函数名
            val func = array[1]
            val map = HashMap<String, Any>()
            player?.let {
                map["player"] = player
                map["papi"] = java.util.function.Function<String, String> { string -> papi(player, string) }
            }
            map["vars"] = java.util.function.Function<String, String> { string -> string.parseSection(cache, player, sections) }
            return ScriptManager.compiledScripts[path]?.invoke(func, map, *args.toTypedArray())?.toString()?.parseSection(parse, cache, player, sections)
        }
        return null
    }
}